pkg/apis/network/v1/egressrouter_types.go

```diff
@@ -0,0 +1,121 @@
+type EgressRouter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec EgressRouterSpec `json:"spec"`
+
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status EgressRouterStatus `json:"status,omitempty"`
+}
+
+// EgressRouterSpec contains the configuration for an egress router.
+// +k8s:openapi-gen=true
+// +kubebuilder:validation:Required
+type EgressRouterSpec struct {
+	// Specification of interface to create/use. e.g; macvlan
+	// +kubebuilder:validation:Required
+	NetworkInterface EgressRouterInterface `json:"networkInterface"`
+
+	// IP configuration arguments.
+	// +kubebuilder:validation:Required
+	IP IPConfig `json:"ip"`
+}
+
+// EgressRouterInterface contains the configuration of interface to create/use.
+// +kubebuilder:validation:Required
+type EgressRouterInterface struct {
+	// Arguments specific to the interfaceType macvlan
+	// +kubebuilder:validation:Required
+	MacVLAN VLANConfig `json:"macVLAN"`
+}
+
+// VLANConfig consists of arguments specific to the EgressRouterInterfaceType
+// +kubebuilder:validation:Required
+type VLANConfig struct {
+	// Usually used for macvlan interface.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Mode string `json:"mode,omitempty"`
+
+	// Name of the master inteface. Usually used for macvlan and ipvlan. Need not be specified if it can be inferred from the IP address.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Master string `json:"master,omitempty"`
+}
+
+// IPConfig contains the configuration for the ip arguments
+// +kubebuilder:validation:Required
+type IPConfig struct {
+	// List of IP addresses to configure on the interface.
+	// +kubebuilder:validation:Required
+	Addresses []EgressRouterAddress `json:"addresses"`
+
+	// List of CIDR blocks that the pod is allowed to connect to via this interface. If not provided, the pod can connect to any destination.
+	// +kubebuilder:validation:Optional
+	// +optional
+	AllowedDestinations []CIDR `json:"allowedDestinations,omitempty"`
+}
+
+// EgressRouterAddress contains a pair of ip cidr and gateway to be configured on the router's interface
+// +kubebuilder:validation:Required
+type EgressRouterAddress struct {
+	// IP is the address to configure on the router's interface. Can be IPv4 or IPv6.
+	// +kubebuilder:validation:Required
+	IP CIDR `json:"ip"`
+	// IP address of the next-hop gateway, if it cannot be automatically determined. Can be IPv4 or IPv6.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+}
+
+// CIDR represents an IPv4 or IPV6 CIDR block.
+type CIDR string
+
+// EgressRouterStatus contains the observed status of EgressRouter. Read-only.
+type EgressRouterStatus struct {
+	// Name of the egress router network
+	// +kubebuilder:validation:Required
+	EgressRouter string `json:"egressRouter,omitempty"`
```

Comment:
Is this necessary?

Reply:
hmm I guess not.. I keep getting confused with this list of routers versus single router.
okay so the EgressRouterStatus object will tell the status of the of the corresponding router in the EgressRouter object. So technically this field is not needed, however I added it for displaying it as a part of the CLI output since we decided not to have the `name` field in the main object and to pick it from the CR's metadata: https://github.com/openshift/cluster-network-operator/pull/852/files#diff-9f879d82ac7bec3c4bd2277d645b009b8c6ca77f034e90d1524a026f9dfe2272R22-R28

We have something similar for egressIPs and egressFirewall so thought of keeping the same format.

Reply:
ok you are right, I don't need this field because Name gets displayed auto on output, my knowledge of k8s api sucks.

 [dsal@bkr-hv02  ~]$ oc get egr  
NAME            EGRESSROUTERS   CONDITION   STATUS 
egress-router